packages/lib-algorand/index.ts

```diff
@@ -0,0 +1,82 @@
+import { DfnsApiClient, DfnsError } from '@dfns/sdk'
+import { GetWalletResponse, GenerateSignatureResponse } from '@dfns/sdk/types/wallets'
+
+import { encodeObj, Transaction, decodeSignedTransaction, encodeUnsignedTransaction, EncodedSignedTransaction } from 'algosdk'
+export type DfnsWalletOptions = {
+  walletId: string
+  dfnsClient: DfnsApiClient
+}
+
+const hexToBuffer = (hex: string): Buffer => {
+  return Buffer.from(hex.replace(/^0x/, ''), 'hex')
+}
+
+type WalletMetadata = GetWalletResponse
+
+const assertSigned = (res: GenerateSignatureResponse) => {
+  if (res.status === 'Failed') {
+    throw new DfnsError(-1, 'signing failed', res)
+  } else if (res.status !== 'Signed') {
+    throw new DfnsError(
+      -1,
+      'cannot complete signing synchronously because this wallet action requires policy approval',
+      res
+    )
+  }
+}
+
+export class DfnsWallet {
+  private readonly dfnsClient: DfnsApiClient
+
+  private constructor(private metadata: WalletMetadata, options: DfnsWalletOptions) {
+    this.dfnsClient = options.dfnsClient
+  }
+
+  public static async init(options: DfnsWalletOptions) {
+    const { walletId, dfnsClient } = options
+    const res = await dfnsClient.wallets.getWallet({ walletId })
+
+    if (res.status !== 'Active') {
+      throw new DfnsError(-1, 'wallet not active', { walletId, status: res.status })
+    }
+
+    const { scheme, curve } = res.signingKey
+    if (scheme !== 'EdDSA') {
+      throw new DfnsError(-1, 'key scheme is not EdDSA', { walletId, scheme })
+    }
+    if (curve !== 'ed25519') {
+      throw new DfnsError(-1, 'key curve is not ed25519', { walletId, curve })
+    }
+
+    if (res.network !== 'Algorand' && res.network !== 'AlgorandTestnet') {
+      throw new DfnsError(-1, 'wallet is not bound to Algorand or AlgorandTestnet', {
+        walletId,
+        network: res.network,
+      })
+    }
+
+    return new DfnsWallet(res, options)
+  }
+
+  public get address(): string {
+    return this.metadata.address!
+
+  }
+
+  public async signTransaction(transaction: Transaction): Promise<string> {
```

Comment:
I chose to return directly the signature because in AlgoSDK we have a function named: `decodeSignedTransaction` that will return a `SignedTransaction` but then the algoSDK doesn't provide an easy way for the user to encode it back to the right format.

They have a dedicated type they internally built: `EncodedSignedTransaction`. However, if we want to return this kind of object, we will need to have a separate logic for multisig and single sig transaction.

So, to KISS, I chose to return the signature (then the user can use helpers provided by algosdk to work with multisig/singlesig transaction)

Reply:
I'm not sure I follow. wouldn't encoding just be [`encoding.encode`](https://github.com/algorand/js-algorand-sdk/blob/develop/src/encoding/encoding.ts#L54)?

one of the core design behind transaction signing is `signedData` field should return a valid transaction that can be broadcasted without any more client side manipulation. I think doing it this way breaks that with multisig. I think the binary payload of the signature call should actually be `EncodedSignedTransaction`, that is the outer most envelope of a broadcastable Algorand transaction.

right now we are just passing in EncodedTransaction, that's not a complete transaction. would prefer the input/output to both be serialized version of `EncodedSignedTransaction`. it would be more consistent in both expectation and behavior whether it's single sig, multsig or in the future maybe logicsig. unless I'm missing something that this cannot be done.

Reply:
It can be done but it will be more complex like I said as the construction of the EncodedSignedTransaction will depend if we sign a musig or not

https://github.com/algorand/js-algorand-sdk/blob/develop/src/multisig.ts#L67
https://github.com/algorand/js-algorand-sdk/blob/develop/src/transaction.ts#L1132

It should be feasible though. I just wanted to avoid extra-complexity

Reply:
it should not depend on multisig, you can turn a transaction into `EncodedSignedTranction` by putting it into the wrapping container.

```
{
   txn: transaction.get_object_for_encoding()
}
```

the [definition](https://github.com/algorand/js-algorand-sdk/blob/6fdfa116f0e47e68982281028fdc817577c6aab7/src/types/transactions/encoded.ts#L383-L408) is independent of multisig.

Reply:
Yes, but the signature location depends on if it's multisig or not

I will do it anyway, I will try to factor as much as possible

Reply:
we should handle that on the server side, ideally in the sdk it should just decode the `signedData`.